src/science/synapse/device.cpp

```diff
@@ -159,6 +159,78 @@ auto Device::uri() const -> const std::string& {
   return uri_;
 }
 
+auto Device::get_logs(
+    const std::string& log_level,
+    std::optional<int64_t> since_ms,
+    std::optional<int64_t> start_time_ns,
+    std::optional<int64_t> end_time_ns,
+    std::optional<std::chrono::milliseconds> timeout) -> std::vector<std::string> {
+
+  synapse::LogQueryRequest request;
+  request.set_min_level(log_level_to_pb(log_level));
+
+  if (since_ms.has_value()) {
+    request.set_since_ms(since_ms.value());
+  } else {
+    if (start_time_ns.has_value()) {
+      request.set_start_time_ns(start_time_ns.value());
+    }
+    if (end_time_ns.has_value()) {
+      request.set_end_time_ns(end_time_ns.value());
+    }
+  }
+
+  grpc::ClientContext context;
+  if (timeout.has_value()) {
+    context.set_deadline(std::chrono::system_clock::now() + timeout.value());
+  }
+
+  synapse::LogQueryResponse response;
+  std::vector<std::string> logs;
+
+  grpc::Status status = rpc_->GetLogs(&context, request, &response);
+
+  if (status.ok()) {
+    for (const auto& entry : response.entries()) {
+      logs.push_back(entry.message());
+    }
+  } else {
+    std::cerr << "Error getting logs: " << status.error_message() << std::endl;
+  }
+
+  return logs;
+}
+
+auto Device::tail_logs(
```

Comment:
I'm realizing that the gRPC client in this device class is sync, when we probably want async (whoops). Ideally a consumer could keep a 'tail_logs' call open & running while they make other device calls, but with this current implementation that's not possible. I'll add a backlog task to change the client from sync to async -- for now, lets not include this